pkg/controllers/workgenerator/controller_test.go

```diff
@@ -143,6 +147,199 @@ func TestGetWorkNamePrefixFromSnapshotName(t *testing.T) {
+					},
+				},
+				Spec: fleetv1beta1.WorkSpec{
+					Workload: fleetv1beta1.WorkloadTemplate{
+						Manifests: []fleetv1beta1.Manifest{{RawExtension: runtime.RawExtension{Raw: []byte("{}")}}},
+					},
+				},
+			},
+			expectChanged: true,
+		},
+		{
+			name: "Do not update the existing work if it already points to the same resource and override snapshots",
+			existingWork: &fleetv1beta1.Work{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      workName,
+					Namespace: namespace,
+					Labels: map[string]string{
+						fleetv1beta1.ParentResourceSnapshotIndexLabel: "1",
+					},
+					Annotations: map[string]string{
+						fleetv1beta1.ParentResourceSnapshotNameAnnotation:                "snapshot-1",
+						fleetv1beta1.ParentClusterResourceOverrideSnapshotHashAnnotation: "hash1",
+						fleetv1beta1.ParentResourceOverrideSnapshotHashAnnotation:        "hash2",
+					},
+				},
+				Spec: fleetv1beta1.WorkSpec{
+					Workload: fleetv1beta1.WorkloadTemplate{
+						Manifests: []fleetv1beta1.Manifest{{RawExtension: runtime.RawExtension{Raw: []byte("{}")}}},
+					},
+				},
+			},
+			expectChanged: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			scheme := serviceScheme(t)
+			objects := []client.Object{resourceSnapshot}
+			if tt.existingWork != nil {
+				objects = append(objects, tt.existingWork)
+			}
+			fakeClient := fake.NewClientBuilder().
+				WithStatusSubresource(objects...).
+				WithScheme(scheme).
+				WithObjects(objects...).
+				Build()
+			// Create reconciler with custom client
+			reconciler := &Reconciler{
+				Client:          fakeClient,
+				recorder:        record.NewFakeRecorder(10),
+				InformerManager: &informer.FakeManager{},
+			}
+			changed, _ := reconciler.upsertWork(ctx, newWork, tt.existingWork, resourceSnapshot)
+			if changed != tt.expectChanged {
+				t.Fatalf("expected changed: %v, got: %v", tt.expectChanged, changed)
+			}
+			upsertedWork := &fleetv1beta1.Work{}
+			if fakeClient.Get(ctx, client.ObjectKeyFromObject(newWork), upsertedWork) != nil {
+				t.Fatalf("failed to get upserted work")
+			}
+			if diff := cmp.Diff(newWork, upsertedWork, cmpOptions...); diff != "" {
+				t.Errorf("upsertWork didn't update the work, mismatch (-want +got):\n%s", diff)
+			}
+			if tt.expectChanged {
+				// check if the deployment is applied
+				var u unstructured.Unstructured
+				if err := u.UnmarshalJSON(upsertedWork.Spec.Workload.Manifests[0].Raw); err != nil {
+					t.Fatalf("Failed to unmarshl the result: %v, want nil", err)
+				}
+				var deployment appsv1.Deployment
+				if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &deployment); err != nil {
+					t.Fatalf("Failed to convert the result to deployment: %v, want nil", err)
+				}
+				if diff := cmp.Diff(testDeployment, deployment); diff != "" {
+					t.Errorf("applyJSONPatchOverride() deployment mismatch (-want, +got):\n%s", diff)
```

Comment:
Hi Ryan! Also a nit about the error message: for this one it might be relevant to the `applyJSONPatchOverride()` function?